neqo-common/src/socket.rs

```diff
@@ -0,0 +1,249 @@
+            // (RFC9000, Section 14).
+            let res = match local_addr {
+                SocketAddr::V4(..) => setsockopt(&s, IpDontFrag, &true),
+                SocketAddr::V6(..) => setsockopt(&s, Ipv6DontFrag, &true),
+            };
+            debug_assert!(res.is_ok());
+            // Request IPv4 type-of-service (TOS) and IPv6 traffic class
+            // information for all incoming packets.
+            let res = match local_addr {
+                SocketAddr::V4(..) => setsockopt(&s, IpRecvTos, &true),
+                SocketAddr::V6(..) => setsockopt(&s, Ipv6RecvTClass, &true),
+            };
+            assert!(res.is_ok());
+            // Request IPv4 time-to-live (TTL) and IPv6 hop count
+            // information for all incoming packets.
+            let res = match local_addr {
+                SocketAddr::V4(..) => setsockopt(&s, IpRecvTtl, &true),
+                SocketAddr::V6(..) => setsockopt(&s, Ipv6RecvHopLimit, &true),
+            };
+            debug_assert!(res.is_ok());
+            s
+        }
+    };
+    Ok(socket)
+}
+
+#[cfg(posix_socket)]
+pub fn emit_datagram_posix<S: AsRawFd>(socket: &S, d: &Datagram) -> usize {
+    let iov = [IoSlice::new(&d[..])];
+    let tos = i32::from(d.tos());
+    let ttl = i32::from(d.ttl());
+    let cmsgs = match d.destination() {
+        SocketAddr::V4(..) => [IpTos(&tos), IpTtl(&ttl)],
+        SocketAddr::V6(..) => [Ipv6TClass(&tos), Ipv6HopLimit(&ttl)],
+    };
+    sendmsg(
+        socket.as_raw_fd(),
+        &iov,
+        &cmsgs,
+        MsgFlags::empty(),
+        Some(&SockaddrStorage::from(d.destination())),
+    )
+    .unwrap()
+}
+
+#[cfg(not(posix_socket))]
+pub fn emit_datagram_generic<S: AsRawFd>(socket: &S, d: &Datagram) -> usize {
+    // Use the default `UdpSocket::send_to` implementation for non-POSIX platforms.
+    // This also means we don't set the TOS and TTL values.
+    sendto(
+        socket.as_raw_fd(),
+        &d[..],
+        &SockaddrStorage::from(d.destination()),
+        MsgFlags::empty(),
+    )
+    .unwrap()
+}
+
+/// Send the UDP datagram on the specified socket.
+///
+/// # Arguments
+///
+/// * `fd` - The UDP socket to send the datagram on.
+/// * `d` - The datagram to send.
+///
+/// # Returns
+///
+/// An `io::Result` indicating whether the datagram was sent successfully.
+///
+/// # Panics
+///
+/// Panics if the send call fails.
+///
+/// # Notes
+///
+/// On non-POSIX platforms, TOS and TTL will not be sent and hence revert to the system default.
+/// TODO: Figure out how to set TOS and TTL at least on Windows.
+pub fn emit_datagram<S: AsRawFd>(socket: &S, d: &Datagram) -> io::Result<()> {
+    #[cfg(posix_socket)]
+    let sent = emit_datagram_posix(socket, d);
+    #[cfg(not(posix_socket))]
+    let sent = emit_datagram_generic(socket, d);
+    if sent != d.len() {
+        eprintln!("Unable to send all {} bytes of datagram", d.len());
+    }
+    Ok(())
+}
+
+// #[cfg(posix_socket)]
```

Comment:
???

Reply:
Leftover debug thing; am still trying to get this to build/run on Windows.